### PR TITLE
Add audformat.utils.replace_file_extension()

### PR DESCRIPTION
**Pull request by hagenw**
Add a utility function to change any existing file extension to the requested one.
As we might need this at different places and you can implement it less efficiently, I thought it would also be good to have it here.

image

### CODE REVIEW

File: tests/test_utils.py [hunk 1]
```diff
@@ -930,6 +930,51 @@ def test_read_csv(csv, result):
...
+        ),
+        (
+            audformat.filewise_index(['f1.WAV', 'f2.WAV']),
+            'MP3',
+            audformat.filewise_index(['f1.MP3', 'f2.MP3']),
+        ),
+        (
+            audformat.filewise_index(['f1', 'f2.wv']),
+            'mp3',
+            audformat.filewise_index(['f1', 'f2.mp3']),
```
Comment (hagenw) [inline tests/test_utils.py:964]: Here I'm not completely sure what should happen for a file name without extension. Should it also be changed to `f1.mp3`?

File: tests/test_utils.py [hunk 2]
```diff
@@ -930,6 +930,51 @@ def test_read_csv(csv, result):
...
+        ),
+        (
+            audformat.filewise_index(['f1.wav', 'f2.wav']),
+            '',
+            audformat.filewise_index(['f1.', 'f2.']),
```
Comment (hagenw) [inline tests/test_utils.py:969]: Here, I'm also not sure if we should return `['f1', 'f2']` instead?

Comment (frankenjoe) [reply tests/test_utils.py:969]: Yes, I think that makes more sense.

File: tests/test_utils.py (hunk 1 above)
Comment (frankenjoe) [reply tests/test_utils.py:964]: > I did not call the function `replace_file_extension()` as we don't ask to replace a particular file extension

But that means I cannot replace `tar.gz` with `zip`, for instance.

**Comment by frankenjoe** [comment]
> I did not call the function `replace_file_extension()` as we don't ask to replace a particular file extension

Which means I cannot replace `tar.gz` with `zip`, for instance.

**Comment by hagenw** [comment]
> > I did not call the function `replace_file_extension()` as we don't ask to replace a particular file extension
> 
> Which means I cannot replace `tar.gz` with `zip`, for instance.

It depends. It's simply up to us how we define the regexp matching pattern to find the file extension. But I think it will be hard to find a pattern that will always work. So at the moment, I would say it will just replace the extension after the last dot.
If I fix what `extension=''` returns, you could replace `tar.gz` by calling `index = set_file_extension(index, '')` and afterwards `index = set_file_extension(index, 'zip')`.

File: tests/test_utils.py (hunk 2 above)
Comment (hagenw) [reply tests/test_utils.py:969]: I changed it and extended the docstring, see the description above.

**Comment by frankenjoe** [comment]
>  It's simply up to us how we define the regexp matching pattern to find the file extension.

Would it maybe make sense that we give the user the option to overwrite the regex pattern?

**Comment by hagenw** [comment]
> > It's simply up to us how we define the regexp matching pattern to find the file extension.
> 
> Would it maybe make sense that we give the user the option to overwrite the regex pattern?

Good idea, this also implicitly documents the used matching pattern. I changed it accordingly.
The only problem is that the formatting in the docstring looks slightly odd:

image

Or is this the correct representation of
```python
r'\.[a-zA-Z0-9]+$'
```
?

**Comment by frankenjoe** [comment]
> The only problem is that the formatting in the docstring looks slightly odd

Maybe use a regular string in the argument and then convert to a raw string later with `pattern = rf'{pattern'}`?

**Comment by hagenw** [comment]
> > The only problem is that the formatting in the docstring looks slightly odd
> 
> Maybe use a regular string in the argument and then convert to a raw string later with `pattern = rf'{pattern}'`?

No, this does not really help. So maybe we just stay with the current solution?

**Comment by hagenw** [comment]
The big problem is that if a user copies the string from the solution and modifies it slightly, e.g. `pattern='\\\\.[A-Z]+$'`, it will fail.

**Comment by frankenjoe** [comment]
> No, this does not really help. So maybe we just stay with the current solution?

Ok, so the problem is really that `sphinx` replaces `\` with `\\\\` correct?

**Comment by hagenw** [comment]
> > No, this does not really help. So maybe we just stay with the current solution?
> 
> Ok, so the problem is really that `sphinx` replaces `\` with `\\\\` correct?

Yes, it replaces first `r'\'` with `'\\'` which is still correct and then `\` with `\\`

**Comment by frankenjoe** [comment]
> Yes, it replaces first `r'\'` with `'\\'` which is still correct and then `\` with `\\`

We could set it to `None` and then say in the docstring what is the default value?

File: audformat/core/utils.py [hunk 3]
```diff
@@ -640,6 +641,62 @@ def same_dtype(d1, d2) -> bool:
...
+        updated index
+
+    Example:
+        >>> idx = filewise_index(['f1.wav', 'f2.flac'])
+        >>> set_file_extension(idx, 'mp3')
+        Index(['f1.mp3', 'f2.mp3'], dtype='object', name='file')
+        >>> idx = filewise_index(['f1.wav.gz', 'f2.wav.gz'])
+        >>> set_file_extension(idx, '')
+        Index(['f1.wav', 'f2.wav'], dtype='object', name='file')
+        >>> set_file_extension(idx, 'flac', pattern='.wav.gz')
```
Comment (frankenjoe) [inline audformat/core/utils.py:675]: For best practise maybe change to ` pattern='\.wav\.gz$'?

File: audformat/core/utils.py [hunk 4]
```diff
@@ -640,6 +641,62 @@ def same_dtype(d1, d2) -> bool:
...
+            the current file extension is removed
+        pattern: regexp pattern to match current extensions.
+            In contrast to ``extension``,
+            you have to include ``'.'``
+
+    Returns:
+        updated index
+
+    Example:
+        >>> idx = filewise_index(['f1.wav', 'f2.flac'])
```
Comment (frankenjoe) [inline audformat/core/utils.py:669]: For consistency with the other examples I would rename to `index`,

Comment (hagenw) [reply audformat/core/utils.py]: There is not really consitency ;)

image

image

image

image

But I agree, that `index` makes the most sense here. I changed it accordingly

Comment (frankenjoe) [reply audformat/core/utils.py]: Agree, `i1` and `i2` is really bad naming. I would like to say it wasn't me who added those examples, but I'm afraid that'd be a lie :)

File: audformat/core/utils.py (hunk 3 above)
Comment (hagenw) [reply audformat/core/utils.py]: Good that you reminded me of that one. I changed to `pattern=r'\.wav\.gz$'`.

Comment (frankenjoe) [reply audformat/core/utils.py]: In principle, the function can now be used to replace any parts of the file path. But I guess it still makes sense to stick keep the name `audformat.utils.set_file_extension()` as this is the most likely use case.

**Comment by hagenw** [comment]
> > Yes, it replaces first `r'\'` with `'\\'` which is still correct and then `\` with `\\`
> 
> We could set it to `None` and then say in the docstring what is the default value?

The problem with using `None` and stating the default string in the documentation is that you might update the default string later in the code and forget about it in the docs. But I also see no other meaningful way at the moment.

File: audformat/core/utils.py (hunk 3 above)
Comment (hagenw) [reply audformat/core/utils.py]: I would now also be fine with using `replace_file_extension()`, but I don't have an opinion which is better.

**Comment by hagenw** [comment]
> > > Yes, it replaces first `r'\'` with `'\\'` which is still correct and then `\` with `\\`
> > 
> > 
> > We could set it to `None` and then say in the docstring what is the default value?
> 
> The problem with using `None` and stating the default string in the documentation is that you might update the default string later in the code and forget about it in the docs. But I also see no other meaningful way at the moment.

I switched now to `pattern=None`.

File: audformat/core/utils.py (hunk 3 above)
Comment (frankenjoe) [reply audformat/core/utils.py]: I think I would prefer `replace_file_extension()` now.

Comment (hagenw) [reply audformat/core/utils.py]: Renamed it

**Pull request merged**